icu4j/main/classes/core/src/com/ibm/icu/impl/breakiter/LSTMBreakEngine.java

```diff
@@ -0,0 +1,444 @@
+            int mat9Size = 4;
+            assert dataLen == mat1Size + mat2Size + mat3Size + mat4Size + mat5Size + mat6Size + mat7Size + mat8Size + mat9Size;
+            int start = 0;
+            this.fEmbedding = make2DArray(data, start, (numIndex+1), embeddings);
+            start += mat1Size;
+            this.fForwardW = make2DArray(data, start, embeddings, 4 * hunits);
+            start += mat2Size;
+            this.fForwardU = make2DArray(data, start, hunits, 4 * hunits);
+            start += mat3Size;
+            this.fForwardB = make1DArray(data, start, 4 * hunits);
+            start += mat4Size;
+            this.fBackwardW = make2DArray(data, start, embeddings, 4 * hunits);
+            start += mat5Size;
+            this.fBackwardU = make2DArray(data, start, hunits, 4 * hunits);
+            start += mat6Size;
+            this.fBackwardB = make1DArray(data, start, 4 * hunits);
+            start += mat7Size;
+            this.fOutputW = make2DArray(data, start, 2 * hunits, 4);
+            start += mat8Size;
+            this.fOutputB = make1DArray(data, start, 4);
+        }
+
+        public EmbeddingType fType;
+        public String fName;
+        public Map<String, Integer> fDict;
+        public float fEmbedding[][];
+        public float fForwardW[][];
+        public float fForwardU[][];
+        public float fForwardB[];
+        public float fBackwardW[][];
+        public float fBackwardU[][];
+        public float fBackwardB[];
+        public float fOutputW[][];
+        public float fOutputB[];
+    }
+
+    // Minimum word size
+    private static final byte MIN_WORD = 2;
+
+    // Minimum number of characters for two words
+    private static final byte MIN_WORD_SPAN = MIN_WORD * 2;
+
+    abstract class Vectorizer {
+        public Vectorizer(Map<String, Integer> dict) {
+            this.fDict = dict;
+        }
+        abstract public void vectorize(CharacterIterator fIter, int rangeStart, int rangeEnd,
+                              List<Integer> offsets, List<Integer> indicies);
+        protected int getIndex(String token) {
+            return fDict.getOrDefault(token, fDict.size());
+        }
+        private Map<String, Integer> fDict;
+    }
+
+    class CodePointsVectorizer extends Vectorizer {
+        public CodePointsVectorizer(Map<String, Integer> dict) {
+            super(dict);
+        }
+
+        public void vectorize(CharacterIterator fIter, int rangeStart, int rangeEnd,
+                              List<Integer> offsets, List<Integer> indicies) {
+            fIter.setIndex(rangeStart);
+            for (char c = fIter.current();
+                 c != CharacterIterator.DONE && fIter.getIndex() < rangeEnd;
+                 c = fIter.next()) {
+                offsets.add(fIter.getIndex());
+                indicies.add(getIndex(String.valueOf(c)));
+            }
+        }
+    }
+
+    class GraphemeClusterVectorizer extends Vectorizer {
```

Comment:
Thought (not actionable): This can probably be optimized at some point to reduce memory allocations (e.g., of the substrings).

Reply:
please articulate your idea. I am not sure sure what you are sugestting here.

Reply:
If you had a CharSequence input instead of CharacterIterator, you could make a subSequence of the CharSequence which is probably more lightweight than your StringBuilder approach.  But, as I said, this requires more code churn and is probably premature optimization.